### ISSUE TITLE
parquet custom dataset 

### ISSUE THREAD

**Issue by rapidrabbit76** [rapidrabbit76/Classification-For-Everyone#10]
```python 

__all__ = ["CustomDataset"]


class CustomDataset(data.Dataset):
    def __init__(
        self,
        parquet_root_path: str,
        trainsforms: Callable,
        columns: List[str] = ["image", "label"],
        num_cached_parquet: int = 5,
        shuffle: Union[int, bool] = True,
    ):
        """parquet data  loader

        Args:
            parquet_root_path (str): parquet 데이터가 존제하는 root dir path
            trainsforms (Callable): 전처리 함수 (Callable!!!)
            columns (List[str], optional): 로더에서 사용할 column list
            num_cached_parquet (int, optional): 캐싱할 parquet 데이터 개수
            shuffle (Union[int, bool], optional): 데이터 셔플링, 테스트시 OFF
        """

        self.shuffle = shuffle
        self.columns = columns
        self.trainsfroms = trainsforms

        self.parquet_list = sorted(
            glob(os.path.join(parquet_root_path, "*.parquet")),
        )
        # parquet file size * num_cached_parquet ~= memory size
        self.num_cached_parquet = num_cached_parquet

        self.steps_cache = int(
            np.ceil(len(self.parquet_list) / self.num_cached_parquet)
        )  # cache step

        self.current_parquet_idx = 0
        self.current_pd_parquets = None  # cached parquets(DataFrame)
        self.current_indices_in_cache = []  # data index in cached parquet
        self.steps_per_epoch = 0
        self.total_len = self.get_total_length()

        self._cache_setting()

    def _download_data(self):
        # TODO: data download(Localy Storage Cached coPy as LSCP)  nfs server to local temploy cache dir
        pass

    def _cache_setting(self):
        """parquet data load from storage (LSCP) to memory"""
        cur_pd, cur_indices = self._cache_parquet(
            self.current_parquet_idx,
        )
        self.current_pd_parquets = cur_pd
        self.current_indices_in_cache = cur_indices

    def get_total_length(self) -> int:
        """read total parquet list and calcurate number of data row

        Returns:
            int: number of dataset's row
        """
        fdf = pq.ParquetDataset(self.parquet_list)
        total_len = 0
        total_len = len(fdf.read(columns=["index"]))
        return total_len

    def _cache_parquet(self, idx: int) -> Tuple[pd.DataFrame, List[int]]:
        """
           parquet data caching
        Args:
            idx (int): index of current parquet gruop state

        Returns:
            Tuple[pd.DataFrame, List[int]]: DataFrame, idx of DF
        """
        next_idx = (idx + 1) * self.num_cached_parquet
        next_idx = None if next_idx > len(self.parquet_list) else next_idx

        # slicing list by cache size
        list_part_parquet = self.parquet_list[idx * self.num_cached_parquet : next_idx]

        # load on memory from LSCP and convert dataframe
        fparquet = pq.ParquetDataset(list_part_parquet)
        df_data = fparquet.read(columns=self.columns).to_pandas()

        # time base seeding
        now = time.time()
        seed = int((now - int(now)) * 100000)
        rng = np.random.RandomState(seed=seed)

        # if shuffle flag Treu DF index shuffle
        np_indices = (
            rng.permutation(len(df_data)) if self.shuffle else np.arange(len(df_data))
        )
        list_indices = np_indices.tolist()

        return df_data, list_indices

    def _dataframe_transfrom(
        self,
        df: pd.DataFrame,
    ) -> Tuple[np.ndarray, int]:
        """data preprocessing

        Args:
            df (pd.DataFrame): dataframe raw

        Returns:
            Tuple[np.ndarray, int]: processed data
        """
        encoded_image = df["image"]
        image = cv2.imdecode(
            np.fromstring(encoded_image, np.byte),
            cv2.IMREAD_COLOR,
        )
        label = df["label"]
        return image, label

    def __len__(self):
        """return total dataset's raw"""
        return self.total_len

    def __getitem__(self, idx):
        # idx는 사용하지 않고 parquet data queue에서 pop
        refresh_idx = 1

        # 현재 캐시 파일 내 데이터를 모두 탐색한 경우
        if len(self.current_indices_in_cache) < refresh_idx:
            # 다음 parquet 파일들을 로딩
            self.current_parquet_idx += 1

            # 모든 parquet 파일 로딩한 경우 첫 번째 파일로
            if self.current_parquet_idx >= self.steps_cache:
                self.current_parquet_idx = 0

            # parquet cache loading
            self._cache_setting()

        # to avoid race conditions, use pop() to get the index
        # race condition to occur by pytorch.data.DataLoader it's use multiprocessing
        pd_idx = self.current_indices_in_cache.pop()
        pd_raw = self.current_pd_parquets.iloc[pd_idx]

        # data processing
        image, label = self._dataframe_transfrom(pd_raw)
        image = self.trainsfroms(image)
        return image, label

```

지금 까지 작업하던 parquet dataset인데 중복 배치 처리랑 race condition 해결해야 될거 같아서 올림 

코드는 네이버에서 작업하던 코드 fastparqeut에서 pyarrow로 변경한거 
네이버코드 설명에 배치중복 처리하는 방식 올려 놓긴 했던데 지금 손이 없어서 올립니다.

구현 예정 기능 

- [ ] 학습 시작할때 NFS에서 로컬로 데이터 가져오는 작업
- [ ] parquet file checksum validation

**Comment by CoLuthien** [comment]
간단히 분석하면 현재 pytorch의 Dataloader 구조는 다음과 같음

1. 각 프로세스마다 하나의 thread unsafe index queue를 지님, 전역 thread process safe data  queue가 존재함
2. 초기화 과정에서 각 프로세스의 인덱스 큐에 인덱스를 할당함. 모든 인덱스는 유니크 함. 
3. dataset이 가진 getitem을 이용해서 데이터를 처리한 뒤 data queue에 삽입함.

즉 현재 데이터 로더는 index data parallel함..

**Comment by rapidrabbit76** [comment]
> 간단히 분석하면 현재 pytorch의 Dataloader 구조는 다음과 같음
> 
> 1. 각 프로세스마다 하나의 thread unsafe index queue를 지님, 전역 thread process safe data  queue가 존재함
> 2. 초기화 과정에서 각 프로세스의 인덱스 큐에 인덱스를 할당함. 모든 인덱스는 유니크 함.
> 3. dataset이 가진 getitem을 이용해서 데이터를 처리한 뒤 data queue에 삽입함.
> 
> 즉 현재 데이터 로더는 index data parallel함..

pytorch dataloader 뜯어보고 다시 말해줄게 고맙습니다. !

**Comment by CoLuthien** [comment]
> > 간단히 분석하면 현재 pytorch의 Dataloader 구조는 다음과 같음
> > 
> > 1. 각 프로세스마다 하나의 thread unsafe index queue를 지님, 전역 thread process safe data  queue가 존재함
> > 2. 초기화 과정에서 각 프로세스의 인덱스 큐에 인덱스를 할당함. 모든 인덱스는 유니크 함.
> > 3. dataset이 가진 getitem을 이용해서 데이터를 처리한 뒤 data queue에 삽입함.
> > 
> > 즉 현재 데이터 로더는 index data parallel함..
> 
> pytorch dataloader 뜯어보고 다시 말해줄게 고맙습니다. !

넵. 일단 요 문제를 해결하려면 IterableDataset을 사용하는 것이 알맞을 듯 합니다.  Map based는 포기해야 할 듯 합니다
자세히는 조금 더 시간이 필요하지만, 일단 이터러블을 사용하는 것으로 각 워커에 나눠주는 인덱스를 마음대로 조정하는 것이 가능할 듯 합니다 :) 
참고: https://pytorch.org/docs/stable/data.html#torch.utils.data.IterableDataset

**Comment by rapidrabbit76** [comment]
> > > 간단히 분석하면 현재 pytorch의 Dataloader 구조는 다음과 같음
> > > 
> > > 1. 각 프로세스마다 하나의 thread unsafe index queue를 지님, 전역 thread process safe data  queue가 존재함
> > > 2. 초기화 과정에서 각 프로세스의 인덱스 큐에 인덱스를 할당함. 모든 인덱스는 유니크 함.
> > > 3. dataset이 가진 getitem을 이용해서 데이터를 처리한 뒤 data queue에 삽입함.
> > > 
> > > 즉 현재 데이터 로더는 index data parallel함..
> > 
> > 
> > pytorch dataloader 뜯어보고 다시 말해줄게 고맙습니다. !
> 
> 넵. 일단 요 문제를 해결하려면 IterableDataset을 사용하는 것이 알맞을 듯 합니다. Map based는 포기해야 할 듯 합니다 자세히는 조금 더 시간이 필요하지만, 일단 이터러블을 사용하는 것으로 각 워커에 나눠주는 인덱스를 마음대로 조정하는 것이 가능할 듯 합니다 :) 참고: https://pytorch.org/docs/stable/data.html#torch.utils.data.IterableDataset

캐싱된 데이터 인덱스를 쪼개서 각 워커에 넣어 주자는 거지? 
Map base라는 말은 뭘까?? map style datasets을 말하는건가?  -> [link](https://pytorch.org/docs/stable/data.html#map-style-datasets)

**Comment by CoLuthien** [comment]
> 

1. 넵 캐싱되는 데이터 인덱스를 각 워커에 규칙을 가지고 쪼개서 넣자는겁니다.
2. 맞습니다. Map style... 용어 조심할게욥..

**Comment by rapidrabbit76** [comment]
> > 
> 
> 1. 넵 캐싱되는 데이터 인덱스를 각 워커에 규칙을 가지고 쪼개서 넣자는겁니다.
> 2. 맞습니다. Map style... 용어 조심할게욥..

ㅇㅋ 주말에 한번 해볼게요 ! 
고마워요! 아니면 만들어서 PR 날려주면 너무 좋고 !

**Comment by CoLuthien** [comment]
> > > 
> > 
> > 
> > 
> > 1. 넵 캐싱되는 데이터 인덱스를 각 워커에 규칙을 가지고 쪼개서 넣자는겁니다.
> > 2. 맞습니다. Map style... 용어 조심할게욥..
> 
> ㅇㅋ 주말에 한번 해볼게요 ! 고마워요! 아니면 만들어서 PR 날려주면 너무 좋고 !

음 시간 남을 때 구현 방향 생각해서 올려두고 방향에 대해서 이야기 해보죠!!

**Comment by CoLuthien** [comment]
갱신
1. torch.utils.data.get_worker_info() 를 사용하면 각 워커의 id를 가져올 수 있음
2. 따라서 Dataset에게 total worker num을 알려준다면, 각 프로세스마다 유일한 parquet cache를  할당하는 것이 가능함
3. 물론 이 방법이 오버헤드가 클 것으로 보임

두 방법 모두 구현해보겠습니당 :)

**Comment by rapidrabbit76** [comment]
이해 하고 있는게 맞는지 검증 
1. 설정한 cache 수 만큼 parquet을 가져오고
2. 각 worker에 cache된 parguet dataframe을 나눠서 worker 마다 유일한 dataframe이 들어가게 한다는거지? 

1에서 메모리로 로드 
2에서 가져온 Dataframe을 나눠준다는거? 

아니면 2에서 메모리 로드작업을 각 워커로 나눠준다는거? 

3. 어떤 오버헤드를 말하는걸까?

**Comment by rapidrabbit76** [comment]
> 갱신
> 
> 1. torch.utils.data.get_worker_info() 를 사용하면 각 워커의 id를 가져올 수 있음
> 2. 따라서 Dataset에게 total worker num을 알려준다면, 각 프로세스마다 유일한 parquet cache를  할당하는 것이 가능함
> 3. 물론 이 방법이 오버헤드가 클 것으로 보임
> 
> 두 방법 모두 구현해보겠습니당 :)

이해 하고 있는게 맞는지 검증

1. 설정한 cache 수 만큼 parquet을 가져오고
2. 각 worker에 cache된 parguet dataframe을 나눠서 worker 마다 유일한 dataframe이 들어가게 한다는거지?

1에서 메모리로 로드
2에서 가져온 Dataframe을 나눠준다는거?

아니면 2에서 메모리 로드작업을 각 워커로 나눠준다는거?

어떤 오버헤드를 말하는걸까?

**Comment by CoLuthien** [comment]
> 

1,2 모두 맞습니당.
근데 제가 이슈를 너무 급하게 갱신한 듯 합니다. 
```python 
import os
import time
from glob import glob
from typing import List, Tuple, Union, Callable

import numpy as np

import torch
from torch.utils import data


class CustomDatasetIter(data.IterableDataset):
    def __init__(self, start, end):
        super(CustomDatasetIter).__init__()

        self.start = start
        self.end = end

    def __iter__(self):
        worker_info = data.get_worker_info()
        n_proc = worker_info.num_workers
        pid = worker_info.id
        

        print(worker_info.dataset)
        return iter(range(pid, self.end, n_proc))

if __name__ == '__main__':
  ds = CustomDatasetIter(0, 32)
  dl = data.DataLoader(ds, num_workers=4, batch_size=4)

  for i in dl:
    print(i)
```

대략 이런 예제를 기반으로 돌려보면 
```
<__mp_main__.CustomDatasetIter object at 0x000001AB68AF7880>
<__mp_main__.CustomDatasetIter object at 0x0000025F01F77880>
<__mp_main__.CustomDatasetIter object at 0x000002A2151E7880>
<__mp_main__.CustomDatasetIter object at 0x0000022CD98A7880>
tensor([ 0,  4,  8, 12])
tensor([ 1,  5,  9, 13])
tensor([ 2,  6, 10, 14])
tensor([ 3,  7, 11, 15])
```
이런 결과를 얻을 수 있고..
문제는 인덱스가 아니라 아마 current_Parquet_idx를 전 프로세스가 공유한다는 가정일 듯 합니다.

**Comment by rapidrabbit76** [comment]
정직하게 뽑아가네

**Comment by CoLuthien** [comment]
> 정직하게 뽑아가네

ㅇㅇ.. 모든 프로세스가 다 따로노는데 인덱스로 놀려니까 먼가 맛이가버린... 
결론적으로 기존코드에서 건드려야할 부분은 cache 프리페치 루틴일것으로 보임니다!!

**Comment by rapidrabbit76** [comment]
> > 정직하게 뽑아가네
> 
> ㅇㅇ.. 모든 프로세스가 다 따로노는데 인덱스로 놀려니까 먼가 맛이가버린... 결론적으로 기존코드에서 건드려야할 부분은 cache 프리페치 루틴일것으로 보임니다!!

ㅇㅇ!! 거기서 나눠야 할거같네 

인덱스 잘 나눠서 넣으면 될거 같은데 ...

**Comment by rapidrabbit76** [comment]
아 맞네 공유하는게 아니라 각 프로세스로 동일하게 복사해서 가지고 있으면 
각자 pop만 하는거니깐 결국 같은 데이터셋으로 계속 돌리고 있는 상황이네 
그러니 학습 배치 마다 거의 같은 데이터셋이 들어가 있어서 loss,acc가 출렁 출렁 하는 상황이고

**Comment by CoLuthien** [comment]
음.. 추가로 구현을 진행하려면 parquet데이터의 스키마가 필요할 듯 합니당...

**Comment by rapidrabbit76** [comment]
@Coluthien

집가면 테스트용 스니핏 만들어 줄게요

**Comment by CoLuthien** [comment]
대략 다음과 같은 형태로 생각하고, cache reading 에서 lock을 사용하는 방식으로 구현하거나..
혹은 iterable을 통해서 각 parquet 파일리스트에 대한 슬라이스를 제공하고
각자 분리된 메모리에서 읽고 쓰는 작업을 하는 방향으로 구현하는게 편할 것 같습니다. 
```python 
global my_list
my_list = mp.Queue(maxsize=0)
class CustomDatasetIter(data.Dataset):
    def __init__(self, start, end, mq):
        super(CustomDatasetIter).__init__()

        self.total_len = end - start
        self.p_list = mq

    def __len__(self):
        """return total dataset's raw"""
        return self.total_len

    def __getitem__(self, index):
        worker_info = data.get_worker_info()
        print (self.p_list.get(), '\n')
        return 1


if __name__ == '__main__':

    for i in range(1, 32):
      my_list.put(i)
    ds = CustomDatasetIter(0, 16, my_list)
    dl = data.DataLoader(ds, num_workers=4, batch_size=4)

    for i, _ in enumerate(dl):
      print()
    
    print("qsize", my_list.qsize())
```

**Comment by rapidrabbit76** [comment]
@CoLuthien 
고맙습니다. 
이 부분은 따로 레포를 파서 진행하자 내일 쉬니깐 진행해볼게!
